src/core/cgroup.c

```diff
@@ -1503,19 +1507,140 @@ static int unit_create_cgroup(
+                return -EINVAL;
+
+        if (!u->manager->system_bus)
+                return -EIO;
+
+        if (!u->cgroup_path)
+                return -EINVAL;
+
+        /* Determine this unit's cgroup path relative to our cgroup root */
+        pp = path_startswith(u->cgroup_path, u->manager->cgroup_root);
+        if (!pp)
+                return -EINVAL;
+
+        if (!isempty(suffix_path))
+                pp = strjoina("/", pp, suffix_path);
+        else
+                pp = strjoina("/", pp);
+
+        path_kill_slashes(pp);
+
+        r = sd_bus_call_method(u->manager->system_bus,
+                               "org.freedesktop.systemd1",
+                               "/org/freedesktop/systemd1",
+                               "org.freedesktop.systemd1.Manager",
+                               "AttachProcessesToUnit",
+                               &error, NULL,
+                               "ssau",
+                               NULL /* empty unit name means client's unit, i.e. us */, pp, 1, (uint32_t) pid);
         if (r < 0)
-                return r;
+                return log_unit_debug_errno(u, r, "Failed to attach unit process via the bus: %s", bus_error_message(&error, r));
+
+        return 0;
+}
+
+int unit_attach_pids_to_cgroup(Unit *u, Set *pids, const char *suffix_path) {
+        CGroupMask delegated_mask;
+        const char *p;
+        Iterator i;
+        void *pidp;
+        int r, q;
+
+        assert(u);
+
+        if (!UNIT_HAS_CGROUP_CONTEXT(u))
+                return -EINVAL;
+
+        if (set_isempty(pids))
+                return 0;
 
-        r = cg_attach_many_everywhere(u->manager->cgroup_supported, u->cgroup_path, u->pids, migrate_callback, u);
+        r = unit_realize_cgroup(u);
         if (r < 0)
                 return r;
 
-        return 0;
+        if (isempty(suffix_path))
+                p = u->cgroup_path;
+        else
+                p = strjoina(u->cgroup_path, "/", suffix_path);
+
+        delegated_mask = unit_get_delegate_mask(u);
+
+        r = 0;
+        SET_FOREACH(pidp, pids, i) {
+                pid_t pid = PTR_TO_PID(pidp);
+                CGroupController c;
+
+                /* First, attach the PID to the main cgroup hierarchy */
+                q = cg_attach(SYSTEMD_CGROUP_CONTROLLER, p, pid);
+                if (IN_SET(q, -EPERM, -EACCES) && MANAGER_IS_USER(u->manager)) {
+                        /* If we are in a user instance, and we can't moe the process ourselves, let's ask the system
+                         * instance about it instead. Since it's more privileged it might be able to move the process
+                         * across the leaves of a subtree who's top node is not owned by us. */
+
+                        q = unit_attach_pid_to_cgroup_via_bus(u, pid, suffix_path);
+                        if (q < 0)
+                                log_unit_debug_errno(u, q, "Going via the bus didn't work either, ignoring: %m");
```

Comment:
This log line will is hard to understand without context. I think it'd be nicer to exchange the order here, so that both errors (with errnos) are logged. This will make things clearer when debugging. Also, I don't like the artificial setting of the errno (EACCESS → EPERM).
```C
 q = cg_attach(SYSTEMD_CGROUP_CONTROLLER, p, pid);
 if (q < 0) {
            log_unit_debug_errno(...);
            if (MANAGER_IS_SYSTEM(u->manager) || !IN_SET(q, -EPERM, -EACCES))
                  continue;
 q = unit_attach_pid_to_cgroup_via_bus(...);
 if (q < 0)
            log_unit_debug_errno(u, q, "Couldn't move process " PID_FMT " to requested cgroup %s via bus: %m", pid, p);
```